scylla/src/transport/load_balancing/default.rs

```diff
@@ -62,17 +68,36 @@ or refrain from preferring datacenters (which may ban all other datacenters, if
             }
         }
         if let Some(ts) = &routing_info.token_with_strategy {
+            // Try to pick some alive local rack random replica.
+            // If preferred rack is not specified, try to pick local DC replica.
+            let local_rack_picked = self.pick_replica(
+                ts,
+                ReplicaLocation::DatacenterAndRack,
+                &self.pick_predicate,
+                cluster,
+            );
+
+            if let Some(alive_local_rack_replica) = local_rack_picked {
+                return Some(alive_local_rack_replica);
+            }
```

Comment:
I'd like to see this fragment of code enclosed in an If statement that would check if a `preferred_rack` field is `Some`.

In its current form, we would unnecessarily iterate through all local replicas in the case when `preferred_rack` is `None`. This is due to the way `self.pick_replica` works if provided with always-false filter.

Reply:
Done, thanks for the hint.